### PR TITLE
feat(backend): Google Calendar sync worker (IND-313)

### PR DESCRIPTION
**Pull request by yanekyuk**
## Summary

- Adds `google_calendar` to the `Toolkit` type union in `IntegrationService`
- Adds `updateIntegrationSyncConfig` and `getActiveIntegrationSyncs` methods to `ChatDatabaseAdapter` for managing the `syncConfig` JSONB column on `network_integrations`
- Creates `IntegrationSyncQueue` — a BullMQ repeatable job that ticks every 5 minutes, finds active google_calendar integrations past their per-row interval, pulls events via Composio `GOOGLESUPER_GOOGLE_CALENDAR_LIST_EVENTS`, and upserts into `networks.metadata.events[]`
- Registers the worker in `main.ts` and adds it to Bull Board
- Adds `PATCH /api/integrations/:toolkit/sync` endpoint to configure sync settings (calendarId, intervalMs, status)
- Adds `google_calendar` to the controller's allowed toolkits list

### Protocol-side changes (network context wiring)

- `IntentIndexer` now accepts an optional `networkContext` parameter and its system prompt includes event-type network awareness (temporal relevance, theme alignment)
- `IntentNetworkGraphFactory` fetches network type/metadata via `getNetwork()` on the evaluated assignment path and passes rendered context to the indexer
- `OpportunityEvaluator` entity-bundle system prompt gains Rule 10 (event network awareness: temporal relevance, theme alignment, co-attendance signal, schedule context)
- `IntentNetworkGraphDatabase` interface updated to include `getNetwork` in its pick list

## Design decisions

- **Owner userId resolution**: `getActiveIntegrationSyncs` joins `network_integrations` with `network_members` to include the network owner's userId, since Composio's `executeToolAction` requires a userId (not a connectedAccountId) to resolve the connected account
- **Deps injection**: The queue follows the established pattern with `IntegrationSyncQueueDeps` for testability — both DB and integration adapters can be injected
- **Event mapping**: Calendar events are mapped to the `events[]` array shape defined by `EventNetworkMetadataSchema` (externalId, title, startTime, endTime, location, description, tags, syncedAt)
- **Time bounds**: Uses the network's `metadata.startDate`/`endDate` as the Google Calendar query window; falls back to now+30d if not set

## Test plan

- [ ] `bunx tsc --noEmit` passes with zero errors
- [ ] `bun run lint` shows zero errors (warnings are pre-existing)
- [ ] Create an event network, link a Google Calendar integration, activate sync via PATCH endpoint
- [ ] Verify worker tick picks up the active sync and populates `metadata.events[]`
- [ ] Verify error handling: sync failure sets `syncConfig.status` to `'error'`
- [ ] Verify IntentIndexer receives rendered network context on evaluated assignments for event-type networks

### CODE REVIEW

File: backend/src/queues/integration.queue.ts [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+      location: item.location ? String(item.location) : undefined,
+      description: item.description ? String(item.description).slice(0, 500) : undefined,
+      tags: [] as string[],
+      syncedAt: new Date().toISOString(),
+    }));
+
+    const updatedMetadata = {
+      ...(meta ?? {}),
+      events: mappedEvents,
+    };
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/queues/integration.queue.ts:225]: Full event replacement is intentional. Google Calendar is the authoritative source — we're treating `metadata.events` as a materialized cache of the external calendar, not as user-editable local state. The sync always replaces the full window to pick up additions, updates, and deletions from Google. Partial merge would require tracking deleted events separately.

File: backend/src/services/integration.service.ts [hunk 2]
```diff
@@ -244,6 +245,47 @@ export class IntegrationService {
...
+   */
+  async getActiveIntegrationSyncs(): Promise<Array<{
+    networkId: string;
+    toolkit: string;
+    connectedAccountId: string;
+    syncConfig: Record<string, unknown>;
+    ownerUserId: string;
+  }>> {
+    return this.db.getActiveIntegrationSyncs();
+  }
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/services/integration.service.ts:306]: `configureSyncConfig` is thin delegation: it validates ownership, looks up the linked integration, strips undefined keys, reads existing config, merges, validates with Zod, and writes. All business logic lives in the adapter methods and the Zod schema — both already exercised by the queue tests and integration tests. Adding unit tests for this would duplicate coverage.

`getActiveIntegrationSyncs` is a single SQL query behind the adapter. Testing it requires a real database, making it an integration test concern, not a unit test.

File: backend/src/queues/integration.queue.ts [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+export class IntegrationSyncQueue {
+  static readonly QUEUE_NAME = QUEUE_NAME;
+
+  readonly queue = QueueFactory.createQueue<IntegrationSyncPayload>(QUEUE_NAME);
+
+  private readonly logger = log.job.from('IntegrationSyncJob');
+  private readonly queueLogger = log.queue.from('IntegrationSyncQueue');
+  private readonly deps: IntegrationSyncQueueDeps | undefined;
+  private worker: ReturnType<typeof QueueFactory.createWorker<IntegrationSyncPayload>> | null = null;
+
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/queues/integration.queue.ts:59]: Added basic queue tests covering the tick handler: interval gating, NaN lastSyncAt handling, and Google Calendar event mapping. The queue class accepts injected deps (SyncDatabaseAdapter, SyncIntegrationAdapter), so the tests run without Redis or Composio.

File: backend/src/queues/integration.queue.ts [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    await dbAdapter.updateIntegrationSyncConfig(
+      networkId,
+      'google_calendar',
+      {
+        ...existingSyncConfig,
+        lastSyncAt: new Date().toISOString(),
+        calendarId,
+        status: 'active',
+      },
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/queues/integration.queue.ts:236]: Fixed. Removed `status: 'active'` from the completion write — the existing status from the spread is preserved, so a user-initiated pause won't be overwritten.

File: backend/src/controllers/integration.controller.ts [hunk 5]
```diff
@@ -146,6 +146,34 @@ export class IntegrationController {
...
+   * PATCH /api/integrations/:toolkit/sync
+   * Body: { networkId: string, calendarId?: string, intervalMs?: number, status?: 'active' | 'paused' }
+   */
+  @Patch('/:toolkit/sync')
+  @UseGuards(RateLimit('write'), AuthGuard)
+  async configureSyncConfig(req: Request, user: AuthenticatedUser, params: { toolkit: string }) {
+    if (!isAllowedToolkit(params.toolkit)) {
+      return new Response(JSON.stringify({ error: 'Unsupported toolkit' }), { status: 400 });
+    }
+    const body = await req.json().catch(() => ({})) as Record<string, unknown>;
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/controllers/integration.controller.ts:160]: Fixed. PATCH sync endpoint now rejects non-`google_calendar` toolkits with a specific error message. Other toolkits can be enabled when their workers are added.

File: backend/src/adapters/database.adapter.ts [hunk 6]
```diff
@@ -3421,6 +3421,124 @@ export class ChatDatabaseAdapter {
...
+        owner.user_id AS owner_user_id
+      FROM network_integrations ni
+      CROSS JOIN LATERAL (
+        SELECT nm.user_id
+        FROM network_members nm
+        WHERE nm.network_id = ni.network_id
+          AND 'owner' = ANY(nm.permissions)
+        LIMIT 1
+      ) owner
+      WHERE ni.sync_config->>'status' = 'active'
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/adapters/database.adapter.ts:3533]: Fixed. Added `JOIN networks n ON ... AND n.deleted_at IS NULL` and `nm.deleted_at IS NULL` filter on the member subquery. Also added `ORDER BY nm.created_at ASC` for deterministic owner selection.

File: backend/src/queues/tests/integration.queue.spec.ts [hunk 7]
```diff
@@ -0,0 +1,274 @@
+/**
+ * Unit tests for IntegrationSyncQueue. Uses injected deps to avoid Redis/Composio/DB.
+ */
+import { config } from 'dotenv';
+config({ path: '.env.test' });
+
+import { describe, expect, it, afterAll, beforeEach } from 'bun:test';
+import { mock } from 'bun:test';
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/queues/tests/integration.queue.spec.ts]: Fixed. Removed unused `beforeEach` import.

File: backend/src/services/integration.service.ts [hunk 8]
```diff
@@ -39,7 +40,7 @@ interface SlackMember {
   profile?: { real_name?: string; email?: string };
 }
 
-type Toolkit = 'gmail' | 'slack';
+type Toolkit = 'gmail' | 'slack' | 'google_calendar';
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/services/integration.service.ts:43]: The \`Toolkit\` type represents all supported integration toolkits across the service — it's used by \`configureSyncConfig\`, \`linkToIndex\`, and \`importContacts\`. Splitting into per-operation types (\`ImportToolkit\`, \`SyncToolkit\`) for a 3-member union adds boilerplate without real safety gain. The explicit \`throw\` at line 109 is the runtime guard, and the controller already blocks \`google_calendar\` from reaching \`importContacts\` at the route level.

File: backend/src/queues/integration.queue.ts [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+
+/**
+ * Integration sync queue: BullMQ repeatable job that ticks every 5 minutes,
+ * finds active integrations past their per-row interval, pulls events via
+ * Composio, and upserts into the network's metadata JSONB.
+ *
+ * @remarks
+ * Workers are started only by the protocol server via {@link IntegrationSyncQueue.startWorker}.
+ * Handlers orchestrate by calling adapters — no business logic here.
+ */
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/queues/integration.queue.ts:49]: Fixed. Updated the docstring to accurately describe the orchestration the class contains.

File: packages/protocol/src/network/indexer/indexer.graph.ts [hunk 10]
```diff
@@ -121,6 +122,17 @@ export class IntentNetworkGraphFactory {
...
+          // Render network context (type, metadata) for the evaluator
+          const network = await this.database.getNetwork(networkId);
+          const renderedContext = network
+            ? renderNetworkContext({
+                type: network.type ?? 'community',
+                title: network.title,
+                prompt: network.prompt,
+                metadata: network.metadata ?? {},
+              })
+            : null;
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply packages/protocol/src/network/indexer/indexer.graph.ts:134]: The \`getNetwork()\` call only runs on the evaluated path — after prompts are confirmed present — and is immediately followed by an LLM round-trip to IntentIndexer (hundreds of ms to seconds). The DB query adds single-digit ms overhead that's invisible next to the LLM call. Combining it into \`getNetworkMemberContext()\` would couple the protocol's database interface to rendering concerns for a micro-optimization that doesn't move the needle.

File: packages/protocol/src/network/indexer/indexer.graph.ts [hunk 11]
```diff
@@ -121,6 +122,17 @@ export class IntentNetworkGraphFactory {
             ? `${intentForIndexing.sourceType}:${intentForIndexing.sourceId ?? ""}`
             : undefined;
 
+          // Render network context (type, metadata) for the evaluator
+          const network = await this.database.getNetwork(networkId);
+          const renderedContext = network
+            ? renderNetworkContext({
+                type: network.type ?? 'community',
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply packages/protocol/src/network/indexer/indexer.graph.ts:129]: Updated PR description with a dedicated section covering the protocol-side changes: IntentIndexer network context param, indexer graph wiring, OpportunityEvaluator Rule 10, and the interface update.

File: backend/src/queues/integration.queue.ts [hunk 12]
```diff
@@ -0,0 +1,246 @@
+import { Job } from 'bullmq';
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/queues/integration.queue.ts]: Fixed. Changed to `import type { Job }` — it's only used as a type annotation.

File: backend/src/queues/integration.queue.ts [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+ */
+export class IntegrationSyncQueue {
+  static readonly QUEUE_NAME = QUEUE_NAME;
+
+  readonly queue = QueueFactory.createQueue<IntegrationSyncPayload>(QUEUE_NAME);
+
+  private readonly logger = log.job.from('IntegrationSyncJob');
+  private readonly queueLogger = log.queue.from('IntegrationSyncQueue');
+  private readonly deps: IntegrationSyncQueueDeps | undefined;
+  private worker: ReturnType<typeof QueueFactory.createWorker<IntegrationSyncPayload>> | null = null;
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/queues/integration.queue.ts:58]: This compiles cleanly on TS 5.9.3. \`ReturnType<typeof QueueFactory.createWorker<IntegrationSyncPayload>>\` uses instantiation expressions, which have been valid since TypeScript 4.7. No change needed.

File: backend/src/queues/integration.queue.ts [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+      );
+
+      if (!result.successful) {
+        throw new Error(`Google Calendar API error: ${result.error}`);
+      }
+
+      const data = typeof result.data === 'string' ? JSON.parse(result.data as string) : result.data;
+      const items = (data?.response_data?.items ?? data?.items ?? []) as Array<Record<string, unknown>>;
+      allEvents.push(...items);
+      pageToken = (data?.response_data?.nextPageToken ?? data?.nextPageToken) as string | undefined;
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/queues/integration.queue.ts]: The \`typeof === 'string'\` check doesn't cause a tsc error (compiles clean), but the \`as string\` cast was a smell. Fixed by assigning \`result.data\` to \`unknown\` first and narrowing from there — removes the cast and makes the defensive guard type-safe. Composio's SDK does return string-encoded JSON in some cases despite the typed interface.

File: backend/src/services/integration.service.ts [hunk 15]
```diff
@@ -244,6 +250,61 @@ export class IntegrationService {
...
+    const integration = linked.find(i => i.toolkit === toolkit);
+    if (!integration) {
+      throw new Error(`No ${toolkit} integration linked to this network`);
+    }
+
+    // Strip undefined keys so we only merge fields the caller actually provided
+    const patch: Record<string, unknown> = {};
+    if (config.calendarId !== undefined) patch.calendarId = config.calendarId;
+    if (config.intervalMs !== undefined) patch.intervalMs = config.intervalMs;
+    if (config.status !== undefined) patch.status = config.status;
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/services/integration.service.ts:280]: Fixed. Added `.min(1)` to `calendarId` in `SyncConfigSchema` — empty strings now fail Zod validation before reaching the worker.

File: backend/src/adapters/database.adapter.ts [hunk 16]
```diff
@@ -3421,6 +3421,127 @@ export class ChatDatabaseAdapter {
...
+      CROSS JOIN LATERAL (
+        SELECT nm.user_id
+        FROM network_members nm
+        WHERE nm.network_id = ni.network_id
+          AND 'owner' = ANY(nm.permissions)
+          AND nm.deleted_at IS NULL
+        ORDER BY nm.created_at ASC
+        LIMIT 1
+      ) owner
+      WHERE ni.sync_config->>'status' = 'active'
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/adapters/database.adapter.ts:3533]: The worker's \`handleTick\` already guards with \`if (sync.toolkit === 'google_calendar')\` at the dispatch level — non-matching rows are skipped with no logs or side effects. Keeping the query generic means adding a new toolkit sync is just an \`else if\` in the worker without touching the adapter layer. The cost of scanning a few extra rows is negligible compared to the Composio API calls.

File: backend/src/queues/integration.queue.ts [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+export interface IntegrationSyncQueueDeps {
+  dbAdapter?: SyncDatabaseAdapter;
+  integrationAdapter?: SyncIntegrationAdapter;
+}
+
+/**
+ * Integration sync queue: BullMQ repeatable job that ticks every 5 minutes,
+ * finds active integrations past their per-row interval, pulls events via
+ * Composio, and upserts into the network's metadata JSONB.
+ *
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/queues/integration.queue.ts:45]: Fixed. Changed "upserts" to "replaces the network's metadata.events array" to match the actual semantics.

File: backend/src/schemas/network.validation.ts [hunk 18]
```diff
@@ -28,7 +28,7 @@ export const NetworkMemberMetadataSchema = z.record(z.string(), z.unknown());
 export const SyncConfigSchema = z.object({
   intervalMs: z.number().min(60_000).optional().default(900_000),
   lastSyncAt: z.string().datetime().optional(),
-  calendarId: z.string().optional().default('primary'),
+  calendarId: z.string().min(1).optional().default('primary'),
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/schemas/network.validation.ts]: Fixed. Added `.trim()` before `.min(1)` so whitespace-only strings are rejected.

File: backend/src/queues/integration.queue.ts [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+    });
+
+    const processor = async (job: Job<IntegrationSyncPayload>) => {
+      this.queueLogger.info(`[IntegrationSyncProcessor] Processing job ${job.id} (${job.name})`);
+      await this.processJob(job.name, job.data);
+    };
+
+    this.worker = QueueFactory.createWorker<IntegrationSyncPayload>(QUEUE_NAME, processor);
+    this.queueLogger.verbose('Integration sync worker started');
+  }
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/queues/integration.queue.ts:89]: This matches the existing queue pattern (\`ProfileQueue.startWorker()\` has the same structure). BullMQ's \`Queue\` handles Redis reconnection internally — if Redis is temporarily unavailable at startup, the connection retries and pending \`add()\` calls complete on reconnect. The \`.catch()\` handles the edge case where the connection is permanently broken. Adding retry logic here would duplicate what BullMQ already provides.

File: backend/src/main.ts [hunk 20]
```diff
@@ -119,6 +120,7 @@ hydeQueue.startCrons();
 emailQueue.startWorker();
 negotiationTimeoutQueue.startWorker();
 negotiationClaimTimeoutQueue.startWorker();
+integrationSyncQueue.startWorker();
```
[the remark this answers is not in the document]
Comment (yanekyuk) [reply backend/src/main.ts:123]: The PR description already has Summary, Protocol-side Changes, Design Decisions, and Test Plan sections — this is a more useful structure for this PR than the changelog format. The changelog format is for merge commit descriptions, not PR bodies during review.
